src/shogun/kernel/PeriodicKernel.cpp

```diff
@@ -0,0 +1,140 @@
+/*
+ * This software is distributed under BSD Clause 3 license (see LICENSE file).
+ *
+ * Written (W) 2015 Esben Soerig
+ */
+
+#include "PeriodicKernel.h"
+
+using namespace shogun;
+
+CPeriodicKernel::CPeriodicKernel() : CDotKernel()
+{
+	init();
+}
+
+CPeriodicKernel::CPeriodicKernel(float64_t ls, float64_t p, int32_t s) : CDotKernel(s)
+{
+	init();
+	set_length_scale(ls);
+	set_period(p);
+}
+
+CPeriodicKernel::CPeriodicKernel(CDotFeatures* l, CDotFeatures* r, float64_t ls,
+								 float64_t p, int32_t s) : CDotKernel(s)
+{
+	init();
+	set_length_scale(ls);
+	set_period(p);
+	init(l,r);
+}
+
+void CPeriodicKernel::precompute_squared_helper(SGVector<float64_t>& buf,
+												CDotFeatures* df)
+{
+	int32_t num_vec=df->get_num_vectors();
+	buf = SGVector<float64_t>(num_vec);
+
+	for (int32_t i=0; i<num_vec; i++)
+		buf[i]=df->dot(i,df, i);
+}
+
+bool CPeriodicKernel::init(CFeatures* l, CFeatures* r)
+{
+	CDotKernel::init(l, r);
+	precompute_squared();
+	return init_normalizer();
+}
+
+float64_t CPeriodicKernel::compute(int32_t idx_a, int32_t idx_b)
+{
+	/* Periodic kernel defined as by David Duvenaud in 
+	 http://mlg.eng.cam.ac.uk/duvenaud/cookbook/index.html 
+
+	 k({\bf x},{\bf x'})= exp(-\frac{2sin^2(|{\bf x}-{\bf x'}|/p)}{l^2})
+	*/
+	float64_t sin_term = CMath::sin(M_PI*distance(idx_a,idx_b)/m_period);
+	return CMath::exp(-2*CMath::pow(sin_term, 2)/CMath::pow(m_length_scale,2));
+}
+
+void CPeriodicKernel::precompute_squared()
+{
+	if (!lhs || !rhs)
+		return;
+
+	CDotFeatures* dotlhs = dynamic_cast<CDotFeatures*>(lhs);
+	REQUIRE(dotlhs != NULL, "Left-hand-side feature matrix could not be cast "
+							"to type CDotFeatures.\n")
+
+	precompute_squared_helper(m_sq_lhs, dotlhs);
+
+	if (lhs==rhs) {
+		m_sq_rhs=m_sq_lhs;
+	}
+	else {
+		CDotFeatures *dotrhs = dynamic_cast<CDotFeatures*>(rhs);
+		REQUIRE(dotrhs != NULL, "Right-hand-side feature matrix could not be "
+								"cast to type CDotFeatures.\n")
+
+		precompute_squared_helper(m_sq_rhs, dotrhs);
+	}
+}
+
+SGMatrix<float64_t> CPeriodicKernel::get_parameter_gradient(
+		const TParameter* param, index_t index)
+{
+	REQUIRE(lhs, "Left-hand-side features not set!\n")
+	REQUIRE(rhs, "Right-hand-side features not set!\n")
+
+	if (!strcmp(param->m_name, "length_scale"))
+	{
+		SGMatrix<float64_t> derivative=SGMatrix<float64_t>(num_lhs, num_rhs);
+
+		for (int j=0; j<num_lhs; j++)
```

Comment:
@sorig Another very annoying minor comment. Although it is perfectly fine, but in shogun, _multi-line_ (not necessarily _multi-statement_) blocks under _if/for/while_  etc usually goes within braces. For single _line_ braces are not use. Brace (both opening and closing) takes an entire new line alone. Also, spaces are not used between operators and operands (I like it personally the other way though).